libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,771 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
-                 " expect: ", beacon_->aeon.members.size() - 1);
+  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to join before starting setup. Connected: ",
+                 connected.size(), " expect: ", beacon_->aeon.members.size() - 1);
 
   return State::WAIT_FOR_DIRECT_CONNECTIONS;
 }
 
-BeaconSetupService::State BeaconSetupService::OnBroadcastID()
+BeaconSetupService::State BeaconSetupService::OnWaitForReadyConnections()
 {
   std::lock_guard<std::mutex> lock(mutex_);
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_READY_CONNECTIONS));
 
-  // Creating identity
-  CabinetMemberDetails member;
-  member.identity = beacon_->manager.identity();
-  member.id       = beacon_->manager.id();
-  member_details_queue_.push_back(member);
+  // Try broadcasting connections again in case they were discarded by RBCs who's cabinet was reset
+  // late
+  fetch::serializers::MsgPackSerializer serializer;
+  serializer << connections_;
+  pre_dkg_rbc_.Broadcast(serializer.data());
 
-  Serializer msgser;
-  msgser << member;
+  if (ready_connections_.size() >= beacon_->aeon.members.size() - 1)
+  {
+    for (auto &m : beacon_->aeon.members)
+    {
+      if (m == identity_)
+      {
+        continue;
+      }
+      if (ready_connections_.find(m.identifier()) != ready_connections_.end())
+      {
+        assert(ready_connections_.at(m.identifier()) == connections_);
+        // TODO(jmw): Strategy if connections for members are different
+      }
+      else
+      {
+        state_machine_->Delay(std::chrono::milliseconds(100));
+        FETCH_LOG_INFO(LOGGING_NAME,
+                       "Waiting for all peers to be ready before starting DKG. Ready: ",
+                       ready_connections_.size(), " expect: ", beacon_->aeon.members.size() - 1);
+        return State::WAIT_FOR_READY_CONNECTIONS;
+      }
+    }
+    FETCH_LOG_INFO(LOGGING_NAME, "All peers connected. Proceeding.");
 
-  // TODO(tfr): Require signed connection
-  endpoint_.Broadcast(SERVICE_DKG, CHANNEL_ID_DISTRIBUTION, msgser.data());
-  return State::WAIT_FOR_IDS;
+    BroadcastShares();
+    return State::WAIT_FOR_SHARE;
+  }
+
+  state_machine_->Delay(std::chrono::milliseconds(100));
+  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to be ready before starting DKG. Ready: ",
+                 ready_connections_.size(), " expect: ", beacon_->aeon.members.size() - 1);
+  return State::WAIT_FOR_READY_CONNECTIONS;
 }
 
-BeaconSetupService::State BeaconSetupService::WaitForIDs()
+BeaconSetupService::State BeaconSetupService::OnWaitForShares()
 {
-  std::lock_guard<std::mutex> lock(mutex_);
-  assert(beacon_ != nullptr);
+  std::unique_lock<std::mutex> lock{mutex_};
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_SHARE));
+
+  if ((coefficients_received_.size() == beacon_->aeon.members.size() - 1) &&
+      (shares_received_.size() == beacon_->aeon.members.size() - 1))
+  {
+    lock.unlock();
```

Comment:
shouldn't this lock protect shares_received.clear() ?

Reply:
Yes, thanks.